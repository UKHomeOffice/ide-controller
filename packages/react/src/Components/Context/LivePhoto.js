// Global imports
import React from 'react';

const LivePhotoContext = React.createContext();

const { Provider: LivePhotoProvider } = LivePhotoContext;
const { Consumer: LivePhotoConsumer } = LivePhotoContext;

export { LivePhotoContext, LivePhotoProvider, LivePhotoConsumer };
