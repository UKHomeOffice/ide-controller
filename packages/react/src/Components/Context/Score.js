// Global imports
import React from 'react';

const ScoreContext = React.createContext();

const { Provider: ScoreProvider } = ScoreContext;
const { Consumer: ScoreConsumer } = ScoreContext;

export { ScoreContext, ScoreProvider, ScoreConsumer };
