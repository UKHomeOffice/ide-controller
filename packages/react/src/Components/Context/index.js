export { EventSourceContext } from './EventSource';
export { LivePhotoContext } from './LivePhoto';
export { ScoreContext } from './Score';
export { StatusContext } from './Status';
