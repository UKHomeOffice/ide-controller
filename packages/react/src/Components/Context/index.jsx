// Global imports
import React from 'react';

const Context = React.createContext();

const { Provider } = Context;
const { Consumer } = Context;

const withContext = (Component) => (props) => (
  // eslint-disable-next-line react/jsx-props-no-spreading
  <Consumer>{(context) => <Component {...props} context={context} />}</Consumer>
);

export { Provider, Consumer, withContext };
