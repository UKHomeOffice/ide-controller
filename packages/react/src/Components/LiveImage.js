import React from 'react'

// Local imports
import './Controller.css';
import Config from './Config'

const LiveImage = () => {
  return (
    <div className="govuk-grid-column-one-third">
      <div className="photoContainer--photo medium at6">
        <img src={ Config.blankAvatar } alt="Live scan" className="responsive" />
      </div>
    </div>
  );
}

export default LiveImage