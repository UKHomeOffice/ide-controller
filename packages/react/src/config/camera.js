export const livePhotoConfig = {
  video: {
    width: 720,
    height: 1080,
    sourceModel: 'C920',
  },
  canvas: {
    width: 300,
    height: 350,
  },
  threshold: 0.9,
  zoomFactor: 1.6,
  imageResolution: 60,
};

export const clearDataTimeout = 900000;

export default livePhotoConfig;
