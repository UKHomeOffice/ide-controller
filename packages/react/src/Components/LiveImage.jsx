// Global imports
import React, { useRef, useEffect, useState } from 'react';
import PropTypes from 'prop-types';

// Local imports
import Video from './Video';
import {
  ResPosenet,
  isBelowThreshold,
  getSourceImageOptions,
} from '../helpers/camera';
import './Controller.css';
import Canvas from './Canvas';
import { livePhotoConfig } from '../config/cameraOptions';

const CAPTURE_OPTIONS = {
  ...livePhotoConfig,
};

const LiveImage = ({ deviceId }) => {
  const canvasRef = useRef('canvas');
  const videoRef = useRef('video');
  const [showVideo, setShowVideo] = useState(true);
  const [showCanvas, setShowCanvas] = useState(false);
  const [sourceImageOptions, setSourceImageOptions] = useState({});

  const estimate = async (net) => {
    if (!videoRef.current) return;
    // parameter(imageSource, imageScaleFactor, flipHorizontal, outputStride)
    const pose = await net.estimateSinglePose(videoRef.current, 0.5, false, 16);
    if (isBelowThreshold(pose)) {
      estimate(net);
    } else {
      videoRef.current.pause();
      const { sx, sy, sWidth, sHeight } = getSourceImageOptions(pose);
      setSourceImageOptions({ sx, sy, sWidth, sHeight });
      setShowCanvas(true);
      setShowVideo(false);
    }
  };

  const initResPosenet = async () => {
    if (!videoRef.current) return;
    const net = await ResPosenet();
    estimate(net);
  };

  useEffect(() => {
    videoRef.current.addEventListener('canplay', initResPosenet);
  }, [videoRef]);

  return (
    <div className="govuk-grid-column-one-third">
      <div className="photoContainer--photo medium at6">
        {showVideo && (
          <Video
            ref={videoRef}
            deviceId={deviceId}
            captureOptions={CAPTURE_OPTIONS}
          />
        )}
        {showCanvas && sourceImageOptions.sx && (
          <Canvas
            sourceImage={videoRef.current}
            sourceImageOptions={sourceImageOptions}
            ref={canvasRef}
            options={CAPTURE_OPTIONS}
          />
        )}
      </div>
    </div>
  );
};

export default LiveImage;

LiveImage.propTypes = {
  deviceId: PropTypes.string,
};

LiveImage.defaultProps = {
  deviceId: null,
};
