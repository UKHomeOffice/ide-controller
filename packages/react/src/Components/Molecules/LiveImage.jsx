// Global imports
import PropTypes from 'prop-types';
import React, { useContext, useEffect, useRef, useState } from 'react';

// Local imports
import { livePhotoConfig } from '../../config/camera';
import {
  getCroppedImageCoordination,
  isGoodPicture,
} from '../../helpers/camera';
import { CanvasImage, CanvasRect, Video } from '../Atoms';
import { LivePhotoContext } from '../Context/LivePhoto';
import { ScoreContext } from '../Context';
import { sendToElectronStore } from '../../helpers/ipcMainEvents';

const LiveImage = ({ cameraId }) => {
  const { setLivePhotoContext } = useContext(LivePhotoContext);
  const { setScoreContext } = useContext(ScoreContext);

  const canvasRef = useRef('canvas');
  const guidCanvasRef = useRef('guidCanvas');
  const videoRef = useRef();

  const [showVideo, setShowVideo] = useState(true);
  const [showCanvas, setShowCanvas] = useState(false);
  const [sourceImageOptions, setSourceImageOptions] = useState({});

  const estimate = async () => {
    const isCameraOffline = !videoRef.current;
    if (isCameraOffline) return;
    const croppedImageCoordination = await getCroppedImageCoordination(
      videoRef.current
    );
    setSourceImageOptions(croppedImageCoordination);
    const isBadQuality = !isGoodPicture(croppedImageCoordination);
    if (isBadQuality) {
      setTimeout(() => estimate(), 50);
    } else {
      sendToElectronStore('Livephoto', 'Taken');
      videoRef.current.pause();
      setShowCanvas(true);
      setShowVideo(false);
      setLivePhotoContext({
        image: canvasRef.current.toDataURL('image/jpeg'),
        timestamp: Date.now(),
      });
    }
  };

  useEffect(() => {
    videoRef.current.addEventListener('canplay', async () => {
      setTimeout(estimate, 1000);
    });
    setScoreContext({});
    sendToElectronStore('Livephoto', 'Initialised');
  }, []);

  return (
    <div className="position-relative">
      {showVideo && (
        <>
          <Video
            ref={videoRef}
            cameraId={cameraId}
            captureOptions={livePhotoConfig}
            className="photoContainer--photo"
          />
          <CanvasRect
            className="photoContainer--photo position-absolute"
            ref={guidCanvasRef}
            width={livePhotoConfig.video.width}
            height={livePhotoConfig.video.height}
            coordinate={{
              x: sourceImageOptions.sourceX,
              y: sourceImageOptions.sourceY,
              width: sourceImageOptions.calculatedWidth,
              height: sourceImageOptions.calculatedHeight,
            }}
          />
        </>
      )}
      {showCanvas && (
        <CanvasImage
          className="photoContainer--photo"
          sourceImage={{
            image: videoRef.current,
            x: sourceImageOptions.sourceX,
            y: sourceImageOptions.sourceY,
            width: sourceImageOptions.calculatedWidth,
            height: sourceImageOptions.calculatedHeight,
          }}
          ref={canvasRef}
          destinationImage={{
            x: 0,
            y: 0,
            width: livePhotoConfig.video.width,
            height: livePhotoConfig.video.height,
          }}
        />
      )}
    </div>
  );
};

export default LiveImage;

LiveImage.propTypes = {
  cameraId: PropTypes.string,
};

LiveImage.defaultProps = {
  cameraId: null,
};
