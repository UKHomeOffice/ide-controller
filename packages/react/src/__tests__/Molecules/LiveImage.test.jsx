// Global imports
import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import 'jest-canvas-mock';

// Local imports
import LiveImage from '../../Components/Molecules/LiveImage';
import { LivePhotoProvider } from '../../Components/Context/LivePhoto';
import { ScoreProvider } from '../../Components/Context/Score';
import { livePhotoConfig } from '../../config/camera';

jest.mock('../../helpers/camera.js', () => ({
  getCroppedImageCoordination: jest.fn(() => ({})),
  isGoodPicture: jest.fn(() => true),
}));

const customRender = (ui) => {
  return render(
    <LivePhotoProvider value={{ setLivePhotoContext: jest.fn() }}>
      <ScoreProvider value={{ setScoreContext: jest.fn() }}>{ui}</ScoreProvider>
    </LivePhotoProvider>
  );
};

describe('<LiveImage />', () => {
  it('renders and match the snapshot', () => {
    const { asFragment } = customRender(<LiveImage cameraId="123" />);

    expect(asFragment).toMatchSnapshot();
  });

  it('Video is shown if showVideo is true', () => {
    customRender(<LiveImage cameraId="123" />);
    const video = screen.getByTestId('atoms-video');

    expect(video).toBeInTheDocument();
  });

  it('CanvasRect is shown if showVideo is true', () => {
    customRender(<LiveImage cameraId="123" />);
    const canvasRect = screen.getByTestId('atoms-canvas-rect');

    expect(canvasRect).toBeInTheDocument();
  });

  it('CanvasImage isnt shown if showCanvas is false', () => {
    customRender(<LiveImage cameraId="123" />);
    const canvasImage = screen.queryByTestId('atoms-canvas-image');

    expect(canvasImage).toBeNull();
  });

  it('Video width prop are equal to livePhotoConfig.video.width', () => {
    customRender(<LiveImage cameraIdId="123" />);
    const video = screen.getByTestId('atoms-video');

    expect(video).toHaveAttribute('width', String(livePhotoConfig.video.width));
  });

  it('Video height prop are equal to livePhotoConfig.video.width', () => {
    customRender(<LiveImage cameraIdId="123" />);
    const video = screen.getByTestId('atoms-video');

    expect(video).toHaveAttribute(
      'height',
      String(livePhotoConfig.video.height)
    );
  });

  it('Video renders properly className', () => {
    customRender(<LiveImage cameraIdId="123" />);
    const video = screen.getByTestId('atoms-video');

    expect(video).toHaveClass('live-image__video');
  });

  it('CanvasRect width prop are equal to livePhotoConfig.video.width', () => {
    customRender(<LiveImage cameraIdId="123" />);
    const rect = screen.getByTestId('atoms-canvas-rect');

    /* Note: useing livePhotoConfig.video.height only because we rotate the camera */
    expect(rect).toHaveAttribute('width', String(livePhotoConfig.video.height));
  });

  it('CanvasRect height prop are equal to livePhotoConfig.video.width', () => {
    customRender(<LiveImage cameraIdId="123" />);
    const rect = screen.getByTestId('atoms-canvas-rect');

    expect(rect).toHaveAttribute('height', String(livePhotoConfig.video.width));
  });

  it('CanvasRect renders properly className', () => {
    customRender(<LiveImage cameraIdId="123" />);
    const video = screen.getByTestId('atoms-canvas-rect');

    expect(video).toHaveClass('live-image__canvas position-absolute');
  });
});
