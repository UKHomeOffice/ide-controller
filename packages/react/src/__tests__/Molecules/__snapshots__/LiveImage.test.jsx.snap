// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LiveImage /> renders and match the snapshot 1`] = `[Function]`;
